client/servers/manager.go

```diff
@@ -0,0 +1,529 @@
+	}
+
+	return copy
+}
+
+// RebalanceServers shuffles the list of servers on this metadata.  The server
+// at the front of the list is selected for the next RPC.  RPC calls that
+// fail for a particular server are rotated to the end of the list.  This
+// method reshuffles the list periodically in order to redistribute work
+// across all known Nomad servers (i.e. guarantee that the order of servers
+// in the server list is not positively correlated with the age of a server
+// in the Nomad cluster).  Periodically shuffling the server list prevents
+// long-lived clients from fixating on long-lived servers.
+//
+// Unhealthy servers are removed when serf notices the server has been
+// deregistered.  Before the newly shuffled server list is saved, the new
+// remote endpoint is tested to ensure its responsive.
+func (m *Manager) RebalanceServers() {
+	// Obtain a copy of the current serverList
+	l := m.getServerList()
+
+	// Shuffle servers so we have a chance of picking a new one.
+	l.shuffleServers()
+
+	// Iterate through the shuffled server list to find an assumed
+	// healthy server.  NOTE: Do not iterate on the list directly because
+	// this loop mutates the server list in-place.
+	var foundHealthyServer bool
+	for i := 0; i < len(l.servers); i++ {
+		// Always test the first server.  Failed servers are cycled
+		// while Serf detects the node has failed.
+		srv := l.servers[0]
+
+		ok, err := m.connPoolPinger.Ping(srv.Addr)
+		if ok {
+			foundHealthyServer = true
+			break
+		}
+		m.logger.Printf(`[DEBUG] manager: pinging server "%s" failed: %s`, srv, err)
+
+		l.cycleServer()
+	}
+
+	// If no healthy servers were found, sleep and wait for Serf to make
+	// the world a happy place again. Update the offline status.
+	if foundHealthyServer {
+		atomic.StoreInt32(&m.offline, 0)
+	} else {
+		atomic.StoreInt32(&m.offline, 1)
+		m.logger.Printf("[DEBUG] manager: No healthy servers during rebalance, aborting")
+		return
+	}
+
+	// Verify that all servers are present
+	if m.reconcileServerList(&l) {
+		m.logger.Printf("[DEBUG] manager: Rebalanced %d servers, next active server is %s", len(l.servers), l.servers[0].String())
+	} else {
+		// reconcileServerList failed because Serf removed the server
+		// that was at the front of the list that had successfully
+		// been Ping'ed.  Between the Ping and reconcile, a Serf
+		// event had shown up removing the node.
+		//
+		// Instead of doing any heroics, "freeze in place" and
+		// continue to use the existing connection until the next
+		// rebalance occurs.
+	}
+
+	return
+}
+
+// reconcileServerList returns true when the first server in serverList exists
+// in the receiver's serverList. If true, the merged serverList is stored as
+// the receiver's serverList. Returns false if the first server does not exist
+// in the list. Newly added servers are appended to the list and other missing
+// servers are removed from the list.
+func (m *Manager) reconcileServerList(l *serverList) bool {
```

Comment:
I do not understand this method at all. The server list we pass in is a potentially shuffled version of the one we're reconciling against here (except for a small window where SetServers could have overwritten one). I think maybe it was useful in Consul which also reconciles servers with Serf? That would make sense.

I'm fairly certain this entire method can be removed and replaced by a call directly to `m.saveServerList(l)` above.

Reply:
Done